src/llama_stack/core/connectors/connectors.py

```diff
@@ -0,0 +1,186 @@
+
+
+class ConnectorServiceConfig(BaseModel):
+    """Configuration for the built-in connector service.
+    :param run_config: Stack run configuration for resolving persistence
+    """
+
+    config: StackConfig
+
+
+async def get_provider_impl(config: ConnectorServiceConfig):
+    """Get the connector service implementation."""
+    impl = ConnectorServiceImpl(config)
+    return impl
+
+
+KEY_PREFIX = "connectors:v1:"
+
+
+class ConnectorServiceImpl(Connectors):
+    """Built-in connector service implementation."""
+
+    def __init__(self, config: ConnectorServiceConfig):
+        self.config = config
+        self.kvstore: KVStore
+
+    def _get_key(self, connector_id: str) -> str:
+        """Get the KVStore key for a connector."""
+        return f"{KEY_PREFIX}{connector_id}"
+
+    async def initialize(self):
+        """Initialize the connector service."""
+
+        # Use connectors store reference from run config
+        connectors_ref = self.config.config.storage.stores.connectors
+        if not connectors_ref:
+            raise ValueError("storage.stores.connectors must be configured in run config")
+        self.kvstore = await kvstore_impl(connectors_ref)
+
+    async def register_connector(
+        self,
+        connector_id: str,
+        connector_type: ConnectorType,
+        source: ConnectorSource,
+        url: str,
+        server_label: str | None = None,
+        server_name: str | None = None,
+        server_description: str | None = None,
+    ) -> Connector:
+        """Register a new connector (idempotent - updates if already exists)."""
+
+        connector = Connector(
+            connector_id=connector_id,
+            connector_type=connector_type,
+            source=source,
+            url=url,
+            server_label=server_label,
+            server_name=server_name,
+            server_description=server_description,
+        )
+
+        key = self._get_key(connector_id)
+        existing_connector_json = await self.kvstore.get(key)
+
+        if existing_connector_json:
+            existing_connector = Connector.model_validate_json(existing_connector_json)
+
+            # Only overwrite if the connector is an exact match; otherwise log and keep existing.
+            if existing_connector.model_dump() != connector.model_dump():
+                logger.info(
+                    "Connector %s already exists with different configuration; skipping overwrite",
+                    connector_id,
+                )
+                return existing_connector
+
+            logger.debug("Connector %s already exists and matches; overwriting with same value", connector_id)
+
+        # Persist full connector, including source (Field is excluded from schema/dumps by default).
+        connector_payload = connector.model_dump()
+        connector_payload["source"] = connector.source
+        await self.kvstore.set(key, json.dumps(connector_payload))
+
+        return connector
```

Comment:
`"idempotent - updates if already exists"`

please self-review this code.

consider -
- why update if the values are the same?
- how will a user know their register attempt failed?